lib/ecto/adapters/mysql/connection.ex

```diff
@@ -11,7 +11,13 @@ if Code.ensure_loaded?(Mariaex.Connection) do
 
     def connect(opts) do
       opts = Keyword.update(opts, :port, @default_port, &normalize_port/1)
-      Mariaex.Connection.start_link(opts)
+      case Mariaex.Connection.start_link(opts) do
+        {:ok, pid} ->
+          repo = Keyword.fetch!(opts, :repo)
+          repo.after_connect(pid)
```

Comment:
I'm not sure we want to promote this anti-pattern:
```elixir
case start_link() do
  {:ok, pid} = ok ->
    do_stuff()
    ok
  {:error, _} = error ->
    error
end
```
If `do_stuff/0` can crash - unless the crash is because `pid` exited - then `pid` becomes orphaned because the parent doesn't know about `pid`. With the current pools, the parent will crash because the workers do not catch or rescue. This means `pid` will get the workers exit signal and exit immediately. However if a connection is started by a `Supervisor`, the crash will be caught and `pid` orphaned. `pid` will remain alive until the `Supervisor` exits. The most common cause of this is for a call in `do_stuff/0` to timeout. The adapters `Tds` and `Mariaex` both make this mistake.

However because the current workers don't catch the crashes, they will exit if an `after_connect/1` crashes. The `Poolboy` pool will instantly(/blindly) restart the workers forever. This means a pool can hit this loop forever, effectively `max_restarts: :infinity`. This may look like fault tolerance but the application is useless because it is just constantly failing, almost "too fault tolerant". If this is happening and it is an anticipated error we want to handle the error closer to where it occurs (as close as possible), otherwise we want the error to bubble up the supervision tree because there is a bug.

If there is a case where errors should not bubble up the supervision tree, the errors should be caught and some form of circuit breaker and/or backoff should be used so that the process doesn't get stuck in a continuous loop of failure.

With the other two pools (`SojournBroker` and `SQL.Sandbox`) the error does bubble up the supervision tree. However if the `after_connect/1` crashes because of a network issue (e.g. database is down) we don't want the error to bubble up, we want to handle it. Using `Postgrex` the `after_connect/1` will crash if the postgres server is unreachable because it's `start_link/1` returns before trying to connect and any queries will fail. Adapters should be sure to block their `connect/1` calls until a connection is established to prevent this.

This leads us back around to having to catch the `connect/1` call in the pools. However if we catch there we face the orphan problem described above.

I'm unsure of the best solution, I know @josevalim said not to build it into the pools. However I think the `after_connect/1` should either be built into the drivers (`Postgrex`, `Mariex` etc.) or the pools. If we are only going to support native queries it might be better to leave this to the drivers to handle appropriately. However if it is to be part of Ecto (so that all adapters can use the feature and to prevent duplication), it would be better to provide a helper function for the pools to call, which calls `connect/1` and `after_connect/1` and ensures no orphaned processes and catches (and logs?!) `after_connect/1` errors.

@JonathanTron, I am sorry for going on for so long. I wanted to highlight some existing issues in the current adapters.